src/backend/session.rs

```diff
@@ -0,0 +1,187 @@
+            // it wants to be notified that the session was closed.
+            let _ = monitor.session_closed(token).await;
+        }
+        // This method intentionally does *not* emit the `Closed` signal.
+        server.remove::<Self, _>(&self.path).await?;
+        Ok(())
+    }
+
+    #[zbus(signal)]
+    async fn closed(signal_emitter: &SignalEmitter<'_>) -> zbus::Result<()>;
+}
+
+#[async_trait]
+/// A trait that backends that create long-lived sessions should implement
+/// to be notified when a session has been closed.
+pub trait SessionImpl: Send + Sync {
+    async fn session_closed(&self, session_token: HandleToken) -> crate::backend::Result<()>;
+}
+
+#[derive(SerializeDict, Type, Debug)]
+#[zvariant(signature = "dict")]
+pub struct CreateSessionResponse {
+    session_id: HandleToken,
+}
+
+impl CreateSessionResponse {
+    pub fn new(token: HandleToken) -> Self {
+        Self { session_id: token }
+    }
+}
+
+#[derive(Default)]
+// Not thread-safe! If it needs to be accessed from several threads,
+// consider wrapping it in a Mutex.
+pub(crate) struct SessionManager {
+    sessions: HashMap<HandleToken, Session>,
+}
+
+impl SessionManager {
+    /// Expects to find a tracked session with the given handle.
+    /// Returns a suitable portal error if not.
+    pub fn try_contains(&self, token: &HandleToken) -> crate::backend::Result<&Session> {
+        self.sessions
+            .get(token)
+            .ok_or(PortalError::NotFound(Self::notfound_msg(token)))
+    }
+
+    /// Tests whether a session with the given handle token already exists.
+    pub fn contains(&self, token: &HandleToken) -> bool {
+        #[cfg(feature = "tracing")]
+        tracing::debug!("SessionManager::contains: tracked sessions: {:?}", &self);
+        self.try_contains(token).is_ok()
+    }
+
+    /// Adds a session to the list of tracked sessions.
+    /// Assumes that no session with the same handle token is already tracked
+    /// (if there was one, it is silently discarded and backends are not
+    /// notified). To avoid such a situation, it is recommended to call
+    /// `contains()` prior to instantiating and adding a session.
+    pub fn add(&mut self, session: Session) {
+        let token = session.token();
+        let _ = self.sessions.insert(token.clone(), session);
+    }
+
+    /// Removes a session from the list of tracked sessions.
+    /// Returns an error if no session with the given handle token was tracked.
+    pub fn remove(&mut self, token: &HandleToken) -> crate::backend::Result<()> {
+        if self.sessions.remove(token).is_some() {
+            Ok(())
+        } else {
+            let message = Self::notfound_msg(token);
+            #[cfg(feature = "tracing")]
+            tracing::error!("{}", message.as_str());
+            Err(PortalError::NotFound(message))
+        }
+    }
+
+    fn notfound_msg(token: &HandleToken) -> String {
+        format!("Unknown session: `{token}`")
+    }
```

Comment:
not sure it is worth a method just for this little error message that is used twice ;-) i would drop it